_posts/2020-10-09-torch-optim/torch_network_with_optim.Rmd

```diff
@@ -0,0 +1,269 @@
+date: 10-01-2020
+categories:
+  - Torch
+  - R
+bibliography: bibliography.bib
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+```{r}
+
+```
+
+This is the fourth and last installment in a series introducing `torch` basics. Initially, we [focused on *tensors*](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch/). To illustrate their power, we coded a complete (if toy-size) neural network from scratch. We didn't make use of any of `torch`'s higher-level capabilities -- not even *autograd*, its automatic-differentiation feature.
+
+This changed in the [follow-up post](https://blogs.rstudio.com/ai/posts/2020-10-05-torch-network-with-autograd). No more thinking about derivatives and the chain rule; a single call to `backward()` did it all.
+
+[Two days ago](https://blogs.rstudio.com/ai/posts/2020-10-07-torch-modules), the code again saw a major simplification. Modules, and compositions thereof, turned out to be much more manageable than tensors, hand-assembled in some DAG[^1].
+
+[^1]: directed acyclic graph
+
+Based on that last state, there are just two more things to do. For one, we still compute the loss by hand. And secondly, even though we get the gradients all nicely computed from *autograd*, we still loop over the model's parameters, updating them all ourselves. You won't be surprised to hear that nothing of this is necessary.
+
+## Losses and loss functions
+
+`torch` comes with all the usual loss functions, such as mean squared error, cross entropy, Kullback-Leibler divergence, and their likes. In general, there are two usage modes.
+
+Taking the example of mean squared error, one way is to call `nnf_mse_loss()` directly on the prediction and ground truth tensors. For example:
+
+```{r}
+x <- torch_randn(c(3, 2, 3))
+y <- torch_zeros(c(3, 2, 3))
+
+nnf_mse_loss(x, y)
+```
+
+    torch_tensor 
+    0.682362
+    [ CPUFloatType{} ]
+
+Other loss functions designed to be called directly start with `nnf_` as well: `nnf_binary_cross_entropy()`, `nnf_nll_loss()`, `nnf_kl_div()` ... and so on. [^2]
+
+[^2]: The prefix `nnf_` was chosen because in PyTorch, the corresponding functions live in [torch.nn.functional](https://pytorch.org/docs/stable/nn.functional.html).
+
+The second way is to define the algorithm in advance and call it at some later time. Here, respective constructors all start with `nn_` and end in `_loss`. For example: `nn_bce_loss()`, `nn_nll_loss(),` `nn_kl_div_loss()` ... [^3]
+
+[^3]: This time, the corresponding PyTorch module is [torch.nn](https://pytorch.org/docs/stable/nn.html).
+
+```{r}
+loss <- nn_mse_loss()
+
+loss(x, y)
+```
+
+    torch_tensor 
+    0.682362
+    [ CPUFloatType{} ]
+
+This method may be preferable when one and the same algorithm, as specified by configurable parameters, should be applied to more than one pair of tensors.
```

Comment:
```suggestion
This method may be preferable when the same algorithm should be applied to more than one pair of tensors, as specified by configurable parameters.
```

Reply:
this one is tricky... if if move that phrase it doesn't quite convey the intention ... clear sign the whole thing is too imprecise